crypto/x509/v3_timespec.c

```diff
@@ -0,0 +1,754 @@
+
+    if (ASN1_BIT_STRING_get_bit(bs, OSSL_TIME_SPEC_BIT_MONTH_JAN)) {
+        if (i > 0 && !BIO_puts(out, ", "))
+            return 0;
+        i++;
+        if (!BIO_puts(out, "JAN"))
+            return 0;
+    }
+    if (ASN1_BIT_STRING_get_bit(bs, OSSL_TIME_SPEC_BIT_MONTH_FEB)) {
+        if (i > 0 && !BIO_puts(out, ", "))
+            return 0;
+        i++;
+        if (!BIO_puts(out, "FEB"))
+            return 0;
+    }
+    if (ASN1_BIT_STRING_get_bit(bs, OSSL_TIME_SPEC_BIT_MONTH_MAR)) {
+        if (i > 0 && !BIO_puts(out, ", "))
+            return 0;
+        i++;
+        if (!BIO_puts(out, "MAR"))
+            return 0;
+    }
+    if (ASN1_BIT_STRING_get_bit(bs, OSSL_TIME_SPEC_BIT_MONTH_APR)) {
+        if (i > 0 && !BIO_puts(out, ", "))
+            return 0;
+        i++;
+        if (!BIO_puts(out, "APR"))
+            return 0;
+    }
+    if (ASN1_BIT_STRING_get_bit(bs, OSSL_TIME_SPEC_BIT_MONTH_MAY)) {
+        if (i > 0 && !BIO_puts(out, ", "))
+            return 0;
+        i++;
+        if (!BIO_puts(out, "MAY"))
+            return 0;
+    }
+    if (ASN1_BIT_STRING_get_bit(bs, OSSL_TIME_SPEC_BIT_MONTH_JUN)) {
+        if (i > 0 && !BIO_puts(out, ", "))
+            return 0;
+        i++;
+        if (!BIO_puts(out, "JUN"))
+            return 0;
+    }
+    if (ASN1_BIT_STRING_get_bit(bs, OSSL_TIME_SPEC_BIT_MONTH_JUL)) {
+        if (i > 0 && !BIO_puts(out, ", "))
+            return 0;
+        i++;
+        if (!BIO_puts(out, "JUL"))
+            return 0;
+    }
+    if (ASN1_BIT_STRING_get_bit(bs, OSSL_TIME_SPEC_BIT_MONTH_AUG)) {
+        if (i > 0 && !BIO_puts(out, ", "))
+            return 0;
+        i++;
+        if (!BIO_puts(out, "AUG"))
+            return 0;
+    }
+    if (ASN1_BIT_STRING_get_bit(bs, OSSL_TIME_SPEC_BIT_MONTH_SEP)) {
+        if (i > 0 && !BIO_puts(out, ", "))
+            return 0;
+        i++;
+        if (!BIO_puts(out, "SEP"))
+            return 0;
+    }
+    if (ASN1_BIT_STRING_get_bit(bs, OSSL_TIME_SPEC_BIT_MONTH_OCT)) {
+        if (i > 0 && !BIO_puts(out, ", "))
+            return 0;
+        i++;
+        if (!BIO_puts(out, "OCT"))
+            return 0;
+    }
+    if (ASN1_BIT_STRING_get_bit(bs, OSSL_TIME_SPEC_BIT_MONTH_NOV)) {
+        if (i > 0 && !BIO_puts(out, ", "))
+            return 0;
+        i++;
+        if (!BIO_puts(out, "NOV"))
+            return 0;
+    }
+    if (ASN1_BIT_STRING_get_bit(bs, OSSL_TIME_SPEC_BIT_MONTH_DEC)) {
+        if (i > 0 && !BIO_puts(out, ", "))
+            return 0;
+        i++;
+        if (!BIO_puts(out, "DEC"))
+            return 0;
+    }
+    return 1;
+}
+
+/*
+ * It might seem like you could just print the bits of the string numerically,
+ * but the fifth bit has the special meaning of "the final week" imputed to it
+ * by the text of ITU-T Recommendation X.520.
+ */
+static int print_bit_week(BIO *out, ASN1_BIT_STRING *bs)
+{
+    int i = 0;
+
+    if (ASN1_BIT_STRING_get_bit(bs, OSSL_TIME_SPEC_BIT_WEEKS_1)) {
```

Comment:
Might it be better to write this as a for loop to de-duplicate all the get_bit and BIO_puts calls, and just skip bit 5 with a continue statement?

Reply:
Done. That was a much more elegant approach! Thank you!